lib/main-loop.js

```diff
@@ -0,0 +1,153 @@
+const {log2Report} = require('../__log-to-report');
+
+
+//
+// Module logging variables (set up when main loop starts and processes input params)
+//
+let scriptLog = null;
+let payloadLog = null;
+
+/**
+ * Implements the main loop for bulk-uploading or bulk-applying changes driven by input CSV file.
+ * 
+ * @param {Object} cliArgs - The arguments provided via command line interface.
+ *                             Parameters are parsed by the 'commander' library.
+ * @param {Object} cliCommand - The commmand object resolved from the command line arguments by `commander` module.
+ * @param {Object} payloadModule - The module that performs the migration operation.
+ *                                  It must export a function named 'payloadFunc_Async'
+ * @param {Object} confirmationRoutinesModule - The module that contains functions to obtain user confirmation.
+ */
+async function loopOverCsvInput_Async (cliArgs, cliCommand, payloadModule, confirmationRoutinesModule) {
+    // Retrieving required command line parameters passed for the invocation
+    const inputCsvFilePath = cliArgs.fromCsvFile;
+    const maxConcurrentUploads = cliArgs.maxConcurrentUploads;
+    const outputFolder = cliArgs.outputFolder;
+
+
+    // Set up logging for the loop
+    const log = logging.setupLogInFolder(outputFolder);
+    scriptLog = log.script;
+    payloadLog = log.payload;
+
+
+    await ensureCloudinaryConfigOrExit_Async();
+
+    const operationOptions = {
+        operation  : cliCommand.name(),
+        dest_cloud : cloudinary.config().cloud_name,
+        parameters : cliArgs,
+    }
+
+    try {
+        await confirmationRoutinesModule.confirmOperationOptionsOrExit_Async(operationOptions);
+    } catch (err) {
+        const msg = 'Migration parameters not confirmed. Terminating';
+        console.error(`🛑 ${msg}`);
+        scriptLog.fatal(operationOptions, msg);
+        // Allowing bunyan to "catch up" on writing the log file: https://github.com/trentm/node-bunyan/issues/37
+        await new Promise(resolve => setTimeout(resolve, 500));
+        process.exit(1);
+    }
+
+    scriptLog.info(operationOptions, 'Migration parameters confirmed. Starting migration routine');
+    const stats = {
+        concurrent: 0,
+        attempted: 0,
+        succeeded: 0,
+        failed: 0
+    }
+
+    console.log('\n\n ~~~~~~~ 🚚  Processing the CSV input ... ~~~~~~~');
+
+    // Initializing visual progress bar
+    await progress.init_Async(inputCsvFilePath);
+
+    // Using async generator to avoid loading the entire input file into memory
+    const inputRecordGeneratorAsync = csvReader.getRecordGenerator_Async(inputCsvFilePath);
+
+    // Using async.mapLimit to limit the number of concurrent operations
+    await async.mapLimit(inputRecordGeneratorAsync, maxConcurrentUploads, async (input) => {
+        let payload = null;
+        let response = null;
+        let summary = {
+            status: 'MIGRATED',
```

Comment:
This would need to be a more involved change. 
The tool had been used with this in place - think it is OK for now.

Created issue to address moving forward: https://github.com/cloudinary/cld-migrate-remote-assets-in-bulk/issues/9
LMK what you think @mgriff 

Cheers!

Reply:
Agreed that it makes sense to move this to it's own issue.